File: frontend/src/types/todo.ts
export type TodoStatus = 'pending' | 'in_progress' | 'completed' | 'expired';
export type TodoVisibility = 'private' | 'public' | 'viewer';
export type TodoType = 'regular' | 'assignment';
export type TodoPublishStatus = 'draft' | 'published';
export type TodoPriority = 'low' | 'mid' | 'high';
export type VisibilityStatus = 'visible' | 'hidden';

export interface Todo {
  id: number;
  owner_id: number;
  title: string;
  description?: string | null;
  notes?: string | null;
  status: TodoStatus;
  priority?: TodoPriority | null;
  // Due datetime in UTC (ISO 8601 string with timezone)
  due_datetime?: string | null;
  completed_at?: string | null;
  expired_at?: string | null;
  is_deleted: boolean;
  deleted_at?: string | null;
  created_at: string;
  updated_at: string;
  is_expired: boolean;
  // Assignment workflow fields
  todo_type?: TodoType;
  publish_status?: TodoPublishStatus;
  assignee_id?: number | null;
  visibility_status?: VisibilityStatus | null;
  assignment_assessment?: string | null;
  assignment_score?: number | null;
  submitted_at?: string | null;
  reviewed_at?: string | null;
  reviewed_by?: number | null;
  is_assignment?: boolean;
  is_published?: boolean;
  is_draft?: boolean;
  can_be_edited_by_assignee?: boolean;
}

export interface TodoListResponse {
  items: Todo[];
  total: number;
}

export interface TodoPayload {
  title: string;
  description?: string | null;
  notes?: string | null;
  priority?: TodoPriority | null;
  // Due datetime in UTC (ISO 8601 string with timezone)
  due_datetime?: string | null;
  status?: TodoStatus;
  todo_type?: TodoType;
  publish_status?: TodoPublishStatus;
  assignee_id?: number | null;
  visibility_status?: VisibilityStatus | null;
}

export interface TodoListParams {
  includeCompleted?: boolean;
  includeDeleted?: boolean;
  status?: TodoStatus | 'pending' | 'in_progress' | 'completed' | 'expired';
  limit?: number;
  offset?: number;
}

// UI-related types
export type ViewFilter = 'all' | 'active' | 'completed' | 'expired' | 'deleted';

export interface TodoItemProps {
  todo: Todo | TodoWithAssignedUser;
  onToggleComplete?: (todo: Todo) => void;
  onEdit: (todo: Todo) => void;
  onComplete?: (todo: Todo) => void;
  onReopen?: (todo: Todo) => void;
  onDelete: (todo: Todo) => void;
  onRestore?: (todo: Todo) => void;
  onViewAttachments?: (todo: Todo) => void;
  attachmentCount?: number;
  totalAttachmentSize?: number;
  loadingId?: number | null;
}

// TaskModal component types
export interface TaskFormState {
  title: string;
  description: string;
  notes: string;
  // Due datetime in local timezone (will be converted to UTC for API)
  dueDate: string;  // Date part in YYYY-MM-DD format
  dueTime?: string;  // Time part in HH:MM format
  priority: TodoPriority;
}

export interface TaskModalProps {
  isOpen: boolean;
  onClose: () => void;
  onSuccess: (todo?: any) => void;
  editingTodo?: Todo | null;
  workflowContext?: boolean;
  onComplete?: (todo: Todo) => void;
  onReopen?: (todo: Todo) => void;
  onDelete?: (todo: Todo) => void;
  onRestore?: (todo: Todo) => void;
  onRequestExtension?: (todo: Todo) => void;
}

// Assignment-related types
export interface AssignableUser {
  id: number;
  email: string;
  first_name: string;
  last_name: string;
  is_active: boolean;
  full_name?: string;
}

export interface TodoWithAssignedUser extends Todo {
  assigned_user?: AssignableUser | null;
  viewers?: TodoViewer[];
}

export interface TodoAssignmentUpdate {
  assigned_user_id?: number | null;
  visibility?: TodoVisibility;
}

export interface TodoViewer {
  id: number;
  user_id: number;
  todo_id: number;
  added_at: string;
  user: AssignableUser;
}

export interface TodoViewersUpdate {
  viewer_ids: number[];
}

// Permission-related types
export interface TodoPermissions {
  canView: boolean;
  canEdit: boolean;
  canDelete: boolean;
  canAssign: boolean;
  canChangeStatus: boolean;
  canAddAttachments: boolean;
  isOwner: boolean;
  isAssignee: boolean;
}

// Assignment component props
export interface UserAssignmentProps {
  todo: Todo | TodoWithAssignedUser;
  assignableUsers: AssignableUser[];
  onAssign: (assignment: TodoAssignmentUpdate) => void;
  onUpdateViewers?: (viewers: TodoViewersUpdate) => void;
  isLoading?: boolean;
}

// Assignment workflow types
export interface AssignmentSubmission {
  notes?: string;
}

export interface AssignmentReview {
  approved: boolean;
  assessment?: string;
  score?: number;
}

export interface AssignmentWorkflowResponse {
  success: boolean;
  message: string;
  todo: Todo;
}

export interface AssignmentStatusProps {
  todo: TodoWithAssignedUser;
  permissions: TodoPermissions;
}

// Extension request types
export type ExtensionRequestStatus = 'pending' | 'approved' | 'rejected';

export interface TodoExtensionValidation {
  can_request_extension: boolean;
  max_allowed_due_date?: string | null;
  days_extension_allowed: number;
  reason?: string | null; // Reason why extension cannot be requested
}

export interface TodoExtensionRequestCreate {
  requested_due_date: string; // ISO date string
  reason: string;
}

export interface TodoExtensionRequestResponse {
  status: ExtensionRequestStatus;
  response_reason?: string;
  new_due_date?: string; // ISO datetime string for date change approval
}

export interface TodoExtensionRequest {
  id: number;
  todo_id: number;
  requested_by_id: number;
  creator_id: number;
  requested_due_date: string;
  reason: string;
  status: ExtensionRequestStatus;
  response_reason?: string;
  responded_at?: string;
  responded_by_id?: number;
  created_at: string;
  updated_at: string;
  // Related objects
  requested_by: AssignableUser;
  creator: AssignableUser;
  responded_by?: AssignableUser;
  todo: Todo;
}

export interface TodoExtensionRequestList {
  items: TodoExtensionRequest[];
  total: number;
  pending_count: number;
  approved_count: number;
  rejected_count: number;
}

// Extension request component props
export interface ExtensionRequestModalProps {
  isOpen: boolean;
  onClose: () => void;
  todo: Todo;
  onSuccess: () => void;
}

export interface ExtensionRequestListProps {
  requests: TodoExtensionRequest[];
  onRespond?: (requestId: number, response: TodoExtensionRequestResponse) => void;
  showActions?: boolean;
}

// Additional component props
export interface AssignedTodosProps {
  onEditTodo?: (todo: Todo) => void;
  onViewAttachments?: (todo: Todo) => void;
}

File: backend/app/schemas/todo.py
from __future__ import annotations

from datetime import UTC, datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field, field_serializer, field_validator

from app.utils.constants import (
    TodoPriority,
    TodoPublishStatus,
    TodoStatus,
    TodoType,
    VisibilityStatus,
)


class TodoBase(BaseModel):
    title: str = Field(..., max_length=255)
    description: Optional[str] = None
    assignee_memo: Optional[str] = None
    priority: Optional[str] = Field(default=TodoPriority.MID.value)
    # Due datetime in UTC (will be converted from user's local timezone on input)
    due_datetime: Optional[datetime] = None
    status: Optional[str] = Field(default=TodoStatus.PENDING.value)
    workflow_id: Optional[int] = None

    # Assignment workflow fields
    todo_type: Optional[str] = Field(default=TodoType.REGULAR.value)
    publish_status: Optional[str] = Field(default=TodoPublishStatus.PUBLISHED.value)
    assignee_id: Optional[int] = None
    visibility_status: Optional[str] = None
    assignment_assessment: Optional[str] = None
    assignment_score: Optional[int] = None

    @field_validator("priority")
    @classmethod
    def validate_priority(cls, value: Optional[str]) -> Optional[str]:
        if value is None:
            return TodoPriority.MID.value
        allowed = {priority.value for priority in TodoPriority}
        if value not in allowed:
            raise ValueError(f"Priority must be one of: {', '.join(sorted(allowed))}")
        return value

    @field_validator("status")
    @classmethod
    def validate_status(cls, value: Optional[str]) -> str:
        if value is None:
            return TodoStatus.PENDING.value
        allowed = {status.value for status in TodoStatus}
        if value not in allowed:
            raise ValueError(f"Status must be one of: {', '.join(sorted(allowed))}")
        return value

    @field_validator("title")
    @classmethod
    def validate_title(cls, value: str) -> str:
        if not value or not value.strip():
            raise ValueError("Title is required")
        return value.strip()

    @field_validator("todo_type")
    @classmethod
    def validate_todo_type(cls, value: Optional[str]) -> str:
        if value is None:
            return TodoType.REGULAR.value
        allowed = {todo_type.value for todo_type in TodoType}
        if value not in allowed:
            raise ValueError(f"Todo type must be one of: {', '.join(sorted(allowed))}")
        return value

    @field_validator("publish_status")
    @classmethod
    def validate_publish_status(cls, value: Optional[str]) -> str:
        if value is None:
            return TodoPublishStatus.PUBLISHED.value
        allowed = {status.value for status in TodoPublishStatus}
        if value not in allowed:
            raise ValueError(
                f"Publish status must be one of: {', '.join(sorted(allowed))}"
            )
        return value

    @field_validator("visibility_status")
    @classmethod
    def validate_visibility_status(cls, value: Optional[str]) -> Optional[str]:
        if value is None:
            return value
        allowed = {status.value for status in VisibilityStatus}
        if value not in allowed:
            raise ValueError(
                f"Visibility status must be one of: {', '.join(sorted(allowed))}"
            )
        return value


class TodoCreate(TodoBase):
    pass


class TodoUpdate(BaseModel):
    title: Optional[str] = Field(default=None, max_length=255)
    description: Optional[str] = None
    assignee_memo: Optional[str] = None
    priority: Optional[str] = None
    # Due datetime in UTC (will be converted from user's local timezone on input)
    due_datetime: Optional[datetime] = None
    status: Optional[str] = None

    # Assignment workflow fields
    todo_type: Optional[str] = None
    publish_status: Optional[str] = None
    assignee_id: Optional[int] = None
    visibility_status: Optional[str] = None
    assignment_assessment: Optional[str] = None
    assignment_score: Optional[int] = None

    @field_validator("priority")
    @classmethod
    def validate_priority(cls, value: Optional[str]) -> Optional[str]:
        if value is None:
            return value
        allowed = {priority.value for priority in TodoPriority}
        if value not in allowed:
            raise ValueError(f"Priority must be one of: {', '.join(sorted(allowed))}")
        return value

    @field_validator("title")
    @classmethod
    def validate_title(cls, value: Optional[str]) -> Optional[str]:
        if value is not None and not value.strip():
            raise ValueError("Title cannot be empty")
        return value.strip() if value else value

    @field_validator("status")
    @classmethod
    def validate_status(cls, value: Optional[str]) -> Optional[str]:
        if value is None:
            return value
        allowed = {status.value for status in TodoStatus}
        if value not in allowed:
            raise ValueError(f"Status must be one of: {', '.join(sorted(allowed))}")
        return value


class TodoRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    owner_id: int
    created_by: Optional[int] = None
    last_updated_by: Optional[int] = None
    title: str
    description: Optional[str] = None
    assignee_memo: Optional[str] = None
    viewer_memo: Optional[str] = None  # Private memo for current viewer
    status: str
    priority: Optional[str] = None
    # Due datetime stored in UTC, serialized with timezone info
    due_datetime: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    expired_at: Optional[datetime] = None
    is_deleted: bool
    deleted_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime
    is_expired: bool

    # Assignment workflow fields
    todo_type: str
    publish_status: str
    assignee_id: Optional[int] = None
    visibility_status: Optional[str] = None
    assignment_assessment: Optional[str] = None
    assignment_score: Optional[int] = None
    submitted_at: Optional[datetime] = None
    reviewed_at: Optional[datetime] = None
    reviewed_by: Optional[int] = None

    @field_serializer(
        "due_datetime",
        "completed_at",
        "expired_at",
        "deleted_at",
        "created_at",
        "updated_at",
        "submitted_at",
        "reviewed_at",
    )
    def serialize_datetime(self, dt: Optional[datetime], _info) -> Optional[str]:
        """Ensure datetime fields are serialized with UTC timezone information."""
        if dt is None:
            return None
        # If datetime is naive, assume it's UTC and add timezone
        if dt.tzinfo is None:
            dt = dt.replace(tzinfo=UTC)
        # Serialize to ISO format
        return dt.isoformat()


class TodoListResponse(BaseModel):
    items: list[TodoRead]
    total: int


class TodoStatusUpdate(BaseModel):
    status: str

    @field_validator("status")
    @classmethod
    def validate_status(cls, value: str) -> str:
        allowed = {status.value for status in TodoStatus}
        if value not in allowed:
            raise ValueError(f"Status must be one of: {', '.join(sorted(allowed))}")
        return value


class TodoExtensionValidation(BaseModel):
    """Validation response for todo extension requests."""

    model_config = ConfigDict(from_attributes=True)

    can_request_extension: bool
    max_allowed_due_date: Optional[datetime] = None
    days_extension_allowed: int = 3
    reason: Optional[str] = None  # Reason why extension cannot be requested

    @field_serializer("max_allowed_due_date")
    def serialize_datetime(self, dt: Optional[datetime], _info) -> Optional[str]:
        """Ensure datetime fields are serialized with UTC timezone information."""
        if dt is None:
            return None
        # If datetime is naive, assume it's UTC and add timezone
        if dt.tzinfo is None:
            dt = dt.replace(tzinfo=UTC)
        # Serialize to ISO format
        return dt.isoformat()


# Assignment workflow schemas
class TodoPublishUpdate(BaseModel):
    """Update publish status of a todo."""

    publish_status: str

    @field_validator("publish_status")
    @classmethod
    def validate_publish_status(cls, value: str) -> str:
        allowed = {status.value for status in TodoPublishStatus}
        if value not in allowed:
            raise ValueError(
                f"Publish status must be one of: {', '.join(sorted(allowed))}"
            )
        return value


class AssignmentSubmission(BaseModel):
    """Submit an assignment for review."""

    assignee_memo: Optional[str] = None  # Optional submission notes


class AssignmentReview(BaseModel):
    """Review an assignment and provide assessment."""

    approved: bool
    assessment: Optional[str] = None
    score: Optional[int] = Field(default=None, ge=0, le=100)


class AssignmentWorkflowResponse(BaseModel):
    """Response for assignment workflow actions."""

    success: bool
    message: str
    todo: Optional[TodoRead] = None


class UserInfo(BaseModel):
    """Basic user information for relationships."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    email: str
    full_name: str
